[Support] Shannon expansion in Boolean Logic library

pcineverdies opened the pull request:
This commit introduces the functionalities to perform the shannon expansion over
a boolean expression, as required by the `fast token delivery` methodology.
The code was first introduced in `5d55122` (by @DanaKossaybati ) and taken almost as
it is.

Jiahui17 commented:
Hi, thank you so much for the contribution!

Before we review the PR in detail, please check out these two documents first (Maybe you are trying to duplicate some existing code in Espresso).

https://en.wikipedia.org/wiki/Boole%27s_expansion_theorem

https://github.com/EPFL-LAP/dynamatic/blob/main/lib/Support/Espresso/cofactor.cpp

pcineverdies commented:
Thanks for your suggestions!
I was not taking into account the existence of some similar functionalities in Espresso.

The advantage of the new files is the integration/coherence with the already-present boolean logic library. 
I will check out what Espresso offers and be back to this PR.

Jiahui17 commented:
> Thanks for your suggestions! I was not taking into account the existence of some similar functionalities in Espresso.

Great, thanks

> The advantage of the new files is the integration/coherence with the already-present boolean logic library. I will check out what Espresso offers and be back to this PR.

Also, feel free to improve the existing code there (remove duplicates as much as possible to make the code clean)

pcineverdies commented:
Sorry for getting back late. 
I recently revised this library; this work is designed for the implementation of the _Fast Token Delivery_ algorithm started by @DanaKossaybati and continued by me. I feel like the Espresso features are not suited enough for such a purpose, taking into account the specific requirements of the algorithm. 

I also plan to PR the rest of the FTD implementation asap (one step at a time though, since it's definitely too large for a single PR). Maybe at that time we might discuss directly about some alternatives.

Jiahui17 commented:
> I feel like the Espresso features are not suited enough for such a purpose, taking into account the specific requirements of the algorithm.

What feature is missing there that you need? Strangely, we included Espresso at some point but we aren't using that at all...

pcineverdies commented:
Espresso is currently employed in `Experimental/lib/BooleanLogic/` (see [here](https://github.com/EPFL-LAP/dynamatic/blob/bdfc973d4ed46ff2b3ac5765c2063543b6425a9a/experimental/lib/Support/BooleanLogic/BoolExpression.cpp#L347)). In the Fast Token Delivery (FTD) algorithm, we need to compute some boolean conditions as sum of products (representing different paths in a CFG to go from node A to node B), and then simplify them through Espresso.

The current design choice in the codebase is not to rely on the Espresso representation of boolean conditions, but rather adopting the objects defined in `BooleanLogic`. This is the reason why it is always necessary to go back and forth from the Espresso representation to the `BooleanLogic` one. Because of this reason, `Shannon.cpp` was designed to directly solve the problem of the shannon expansion by relying only on `BooleanLogic`. 
I guess that the design choice comes from the fact that Espresso is mainly about minimization of boolean conditions. Do you feel like this requires a change in the design?

Jiahui17 reviewed (changes requested):
@pcineverdies thanks again for the contributions!

I just left a few comments (mainly on terminologies). One major thing to think about is whether we use SOP or a tree of expressions for storing boolean expressions. Is it easier to generate SOPs or a tree of expressions in the fast token delivery code? 

Also, maybe the code gets cleaner if we generate BDD (i.e., the mux tree in your terminology) from SOPs.

pcineverdies commented on experimental/include/experimental/Support/BooleanLogic/Shannon.h:
```diff
@@ -0,0 +1,86 @@
+namespace experimental {
+namespace boolean {
+
+/// Forward declaration of the Multiplexer struct.
+struct Multiplexer;
+
+/// Represents either a boolean expression or a Multiplexer.
+struct MultiplexerIn {
+  Multiplexer *mux;
+  std::optional<BoolExpression *> boolexpression;
+
+  /// Constructor for initializing with a Multiplexer.
+  MultiplexerIn(Multiplexer *m) : mux(m) {}
+
+  /// Constructor for initializing with a boolean expression.
+  MultiplexerIn(BoolExpression *b) : mux(nullptr), boolexpression(b) {}
+
+  /// Prints the content of the MultiplexerIn based on its type.
+  void print();
+};
+
+/// Represents a multiplexer for boolean logic expressions.
+struct Multiplexer {
+  // Input 0 for the Multiplexer.
+  MultiplexerIn *in0;
+  // Input 1 for the Multiplexer.
+  MultiplexerIn *in1;
+  // Condition for selecting the input.
+  MultiplexerIn *cond;
+
+  /// Constructor for initializing the Multiplexer with inputs and condition.
+  Multiplexer(MultiplexerIn *input0, MultiplexerIn *input1,
+              MultiplexerIn *condition)
+      : in0(input0), in1(input1), cond(condition) {}
+
+  /// Prints the content of the Multiplexer.
+  void print();
+};
```
I agree on the idea of merging the two structures.
Also, I really appreciate your reference to BDDs. This makes way more sense in a boolean logic library rather than `Multiplexer` (that is an hardware resource).
I am going to rename everything accordingly.

pcineverdies commented on experimental/include/experimental/Support/BooleanLogic/Shannon.h:
```diff
@@ -0,0 +1,86 @@
+  void print();
+};
+
+/// Replaces a variable with a specified value in a boolean expression.
+void replaceVarWithValue(BoolExpression *exp, const std::string &var,
+                         ExpressionType t);
```
Agree, fixed.

pcineverdies commented on experimental/include/experimental/Support/BooleanLogic/Shannon.h:
```diff
@@ -0,0 +1,86 @@
+BoolExpression *shannonExpansionNegative(BoolExpression *exp,
+                                         const std::string &var);
+
+/// Applies Shannon decomposition on a boolean expression based on a list of
+/// cofactors.
+/// Returns a dynamically-allocated variable
+MultiplexerIn *applyShannon(BoolExpression *exp,
+                            const std::vector<std::string> &cofactorList);
```
Fixed

pcineverdies commented:
Thanks for your review!
I addressed the concerns about the naming choice and the adopted data structures. 

From my own point of view, it might be worth to rethink about the boolean logic library (as discussed in #177) by using an external dependency to manipulate boolean expression and to simplify them (getting rid of Espresso). 
However, this might require a bit of refactoring on the FTD side, which is not in our priorities.

My plan is to consider such a change once that the whole FTD flow has been merged (maybe by having a separate discussion in which we evaluate possible FOSS alternatives depending on the different requirements from the community).

Jiahui17 commented on experimental/include/experimental/Support/BooleanLogic/BDD.h:
```diff
@@ -0,0 +1,54 @@
+/// variable and possibly 2 subsequent nodes, stored into `inputs`
+struct BDD {
+  // First element is the `negative` branch; second is the `positive` branch
+  std::optional<std::pair<BDD *, BDD *>> inputs;
```
Aren't these `outputs` or `successors`?

pcineverdies replied:
Right, that was still the mux jargon.

Jiahui17 reviewed (approved):
Everything looks good :) Please change the variable name before merging.

Maybe we could look together at how you will use the Boolean expression in the next PR to see how to change the Boolean library.

pcineverdies commented:
Thanks for the review again!
I agree on having a discussion about the way we manipulate boolean conditions.
We will handle the problem on a separate issue - considering both our needs and someone else's needs.
In the meanwhile, I will collect some alternatives!
